### ISSUE TITLE
Consistency of search methods with start position

### ISSUE THREAD

**Issue by ScottPJones** [JuliaLang/julia#12906]
`search("foobar","o",0)` gives a `BoundsError`, but `search("foobar","oo",0)` returns `2:3`.
`search("foobar","o",7)` returns `0:-1`, even though 7 is out of bounds of the string `"foobar"`
`search("foobar","o",8)` gives a `BoundsError`, but `search("foobar","oo",8) returns `0:-1`

`rsearch("foobar","o",0)` returns `0:-1`, even though 0 is out of bounds of the string `"foobar"`
`rsearch("foobar","o",8)` gives a `BoundsError`, but `search("foobar","oo",8) returns `2:3`

My questions are: should any start value which is out of bounds be allowed to return a value?
It seems like 0 and 7 should always get a BoundsError for "foobar".

`search("foobar", Char[], start)` returns a scalar value instead of a range, giving BoundsError if start < 1 or > sizeof(str)+1.  (again, the +1 seems problematic).

**Comment by eschnett** [comment]
With six letters, a search can meaningfully start at seven positions: before each of the six letters, and past the last letter. Thus `search(..., 7)` should not lead to a `BoundsError`. The case of `rsearch(..., 0)` is similar.

**Comment by ScottPJones** [comment]
How can a search start outside the bounds?  Can you give some case where that makes sense?  Where it could make code simpler or cleaner?
What about the inconsistencies between searching for "o" and "oo" in the above examples?
If "o" gets an `BoundsError`, so should "oo".
If you have a loop using `search`, and it gets to the end without finding the string, it will return `0:-1`.  Why does it need to accept something *past* the end?
Do other languages allow starting past the end? (languages that give any errors at all, that is)

**Comment by nalimilan** [comment]
I agree that raising a `BoundsError` when passing indexes below 1 and beyond the last index would be better. That would allow catching bugs, and would be consistent with how indexing strings works.

**Comment by eschnett** [comment]
Things are often cleaner if one allows for the zero case. 

For example, the empty string `""` can be found after the end of other strings. Or, if you are looking for something in an empty string, the only possible starting position is past the end.

**Comment by eschnett** [comment]
Taking substrings of a string also allows starting one-past-the end, if you take a substring of length zero.

The "below 1" case is for backwards searching; it's the same argument, just one-before-the-beginning instead of one-past-the end.

**Comment by ScottPJones** [comment]
In other languages (which I've used, and ones that I've implemented) it was never necessary to allow out-of-bounds start values, so I don't see why it is good in Julia.  Do you have some code that shows a use case where it is simpler?

**Comment by eschnett** [comment]
C and C++ allow "out-of-bounds positions".

Note: When we talk about "out-of-bounds" here, we mean "just past the end", or "just before the beginning". In C, this would mean that your initial position points to the trailing `NUL` character of the string (i.e. "past the end of the string"). This is definitively okay.

Julia is doing everything right.

**Comment by ScottPJones** [comment]
No, C/C++ don't do any bounds checking, period.
There is no guarantee that there is a \0 after a string either, that is just a convention, supported by literals and the old str* functions (which are not recommended to be used any longer anyway).

**Comment by nalimilan** [comment]
What do other high-level languages do?

@eschnett Even if we allow an index of zero or one position beyond the end of the string, why should any out of bounds index be accepted? What's the use case?

**Comment by eschnett** [comment]
The C/C++ standards say clear which cases are allowed and which one aren't. And yes, having a trailing `NUL` is still very much standard in C, and the `str...` functions are in no way deprecated.

But let's move to C++ and the `string` class; see e.g. <http://en.cppreference.com/w/cpp/string/basic_string/substr> for C++: The case `pos==size()` is allowed; given that C++ starts counting from zero, this points past the end of the string.

Oh! Wait! What does this page say? There's an "out_of_range" exception if the position is wrong? So much for "C++ doesn't do any checking"...

**Comment by eschnett** [comment]
@nalimilan The use case is empty strings. If you want to look for an empty string, you can indeed find it at the end of other strings. For example, `"abc"` contains the empty string at positions 1, 2, 3, and 4.

You can also look in the empty string. (Of course, you can only find an empty string there, nothing else.) This empty string is located at position 1 -- which is past the end of the string.

I'm not saying we should allow all out-of-bounds positions. But when looking in a string of length `n`, then there are `n+1` starting positions, ranging from `1` to `n+1`.

**Comment by ScottPJones** [comment]
In C/C++, which are 0 based, the equivalent to a 0 in julia would be -1, which is not allowed.

>I'm not saying we should allow all out-of-bounds positions. But when looking in a string of length n, then there are n+1 starting positions, ranging from 1 to n+1.

It seems you agree here, that 0 is not a valid starting position.

**Comment by eschnett** [comment]
There are two cases, `search` and `rsearch`.

For `search`, valid starting positions are `1` to `n+1`.
For `rsearch`, valid starting positions are `0` to `n`. 
Everything else should be rejected.

And I just found this:
```
search("abc","ab",-1)
LoadError: BoundsError: attempt to access 3-element Array{UInt8,1}:
 0x61
 0x62
 0x63
  at index [0]
while loading In[126], in expression starting on line 1

 in _searchindex at /Users/eschnett/julia/lib/julia/sys.dylib
 in searchindex at /Users/eschnett/julia/lib/julia/sys.dylib
 in search at /Users/eschnett/julia/lib/julia/sys.dylib
```
A bounds check is missing somewhere.

Apologies for being dense earlier; not all the cases that are listed above should be rejected, but some of them should. In particular, `"o"` and `"oo"` should be treated the same.

**Comment by ScottPJones** [comment]
OK, that's the sort of information that I needed.  Thanks!
I'd like to make these all consistent, and make sure they have good unit tests.

**Comment by eschnett** [comment]
@ScottPJones It seems the code in `search.jl` could use a bit of cleanup. For example, it isn't clear to me what the internal functions `_searchindex` and `searchindex` should do on index errors -- return `0` or throw? Some do throw, but only in certain cases.

Also, it seems these functions are not type-stable. If the third argument `i` is present but is not of type `Int`, the local variables will change type, and the return type will also vary. The easiest way out os probably to define the algorithm only for `start::Int`, and to define one additional, generic function that calls `Int()` and forwards. Otherwise, one would need to add many calls to convert, for every call to `next` or `length` or `endof`.

Finally, the order in which these functions are defined seems random. I would sort them either by type (string, array, etc.) so that equivalent `search` and `rsearch` are kept together, or have all `search` methods first, and then all `rsearch` methods. (I'd probably go for the former.)

The return values for searching empty strings or zero-length arrays seem inconsistent; they are sometimes reported as "found", and other times as "not found".

`Chars` can be a vector or a set of characters, but not a tuple. That seems strange; presumably, the expression `('a','b')` is more efficient (requires no heap allocation) than `['a','b']`. The operator `in` does accept tuples, so this seems like an oversight.

**Comment by ScottPJones** [comment]
@eschnett Yes, that's why I posted this issue, I realized that it needed cleaning up, but wasn't sure exactly *what* was the correct result in some of the cases, since they were all inconsistent.
Thanks for the input!
After I finish coverage testing for `search.jl`, and after 0.4 branch, I'll take a stab at cleaning it up.

**Comment by nalimilan** [comment]
@StefanKarpinski @bkamins Can we close this now? Of course `search` no longer exists, but I wonder whether its `find*` replacement now behave as expected.

**Comment by bkamins** [comment]
I have made an initial investigation and have found a following inconsistency:
```
julia> Base.findprev(GenericString(""), GenericString("1234"), 0)
ERROR: BoundsError: attempt to access "1234"
  at index [0]
Stacktrace:
 [1] _rsearchindex(::GenericString, ::GenericString, ::Int64) at .\strings\search.jl:304
 [2] _rsearch(::GenericString, ::GenericString, ::Int64) at .\strings\search.jl:414
 [3] findprev(::GenericString, ::GenericString, ::Int64) at .\strings\search.jl:448
 [4] top-level scope

julia> Base.findprev("", "1234", 0)
1:0
```
At its core it is related to bounds of `thisind` (if we allow `0` and `ncodeinits(n)+1` as valid arguments passed to `thisind`). See the discussion in #25478.
